### PR TITLE
Fix/tg 216 event emitter memory leak

### PR DESCRIPTION
**Pull request by borys-ihnatyev**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/garage-it/SmartHouse-backend/pull/57" target="_blank">ReviewNinja

File: src/mqtt-client/client.js [hunk 1]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
+            event: 'device-info',
+            device: deviceName,
+            value: message
+        };
+    }
+    else {
+        event = {
+            event: 'status',
+            device: deviceName,
+            value: rawMessage.toString()
```
Comment (yakimko) [inline src/mqtt-client/client.js]: mb value: message?

File: src/mqtt-client/client.js [hunk 2]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
-        } else  if (event.event === DEVICE_INFO_EVENT) {
+function onMqttMessage(topic, rawMessage) {
+    debug(`MQTT >>. Got message: topic '${topic}', message: '${rawMessage.toString()}'`);
+    let message = '';
+    try {
+        message = JSON.parse(rawMessage);
+    } catch (e) {
+        message = rawMessage.toString();
+    }
+    let isDeviceInfo = typeof message === 'object';
```
Comment (yakimko) [inline src/mqtt-client/client.js]: const isDeviceInfo

File: src/mqtt-client/client.js [hunk 3]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
-            message = rawMessage.toString();
-        }
-        let isDeviceInfo = typeof message === 'object';
-        let event,
-            deviceName = topic.split('/').pop();
+function onInputEvent(event) {
+    let message, topic = MQTT_OUTPUT_TOPIC_PREFIX;
 
-        if (isDeviceInfo) {
+    if ( event.event === DEVICE_STATUS_EVENT) {
```
Comment (yakimko) [inline src/mqtt-client/client.js]: please remove the space before event

File: src/mqtt-client/client.js [hunk 4]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
-            return;
-        }
+        publishingQueue.delete(deviceName);
+        blockPublishing = false;
+        publishDeviceInfoEvent();
 
-        client.publish(topic, message);
-    });
+        event  = {
+            event: 'device-info',
```
Comment (ashilay) [inline src/mqtt-client/client.js]: DEVICE_INFO_EVENT

File: src/mqtt-client/client.js [hunk 5]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
+}
 
-        if ( event.event === DEVICE_STATUS_EVENT) {
-            message = event.value;
-            topic += event.device;
-        } else  if (event.event === DEVICE_INFO_EVENT) {
+function onMqttMessage(topic, rawMessage) {
+    debug(`MQTT >>. Got message: topic '${topic}', message: '${rawMessage.toString()}'`);
+    let message = '';
+    try {
```
Comment (ashilay) [inline src/mqtt-client/client.js]: Better to create private method, smth like parseMessage()

File: src/mqtt-client/client.js (hunk 4 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: I just extracted this code into function, I was not intend to refactor all this stuff around

File: src/mqtt-client/client.js [hunk 6]
```diff
@@ -28,66 +28,66 @@ const client = mqtt.connect({
...
-    });
+        event  = {
+            event: 'device-info',
+            device: deviceName,
+            value: message
+        };
+    }
+    else {
+        event = {
+            event: 'status',
```
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: I just extracted this code into function, I was not intend to refactor all this stuff around

File: src/mqtt-client/client.js (hunk 3 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: I just extracted this code into function, I was not intend to refactor all this stuff around

File: src/mqtt-client/client.js (hunk 5 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: !fixed

File: src/mqtt-client/client.js (hunk 4 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: !fixed

File: src/mqtt-client/client.js (hunk 1 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: !fixed

File: src/mqtt-client/client.js (hunk 3 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: !fixed

File: src/mqtt-client/client.js [hunk 7]
```diff
@@ -25,93 +26,59 @@ const client = mqtt.connect({
     auth: `${config.mqtt.username}:${config.mqtt.password}`
 });
 
-let blockPublishing = false,
-    publishingQueue = new Set();
+const deviceInfoPublishQueue = uniqueQueue();
+const deviceInfoPublishQueueLock = lock();
+let deviceInfoPublishQueueClearTimeout = 0;
```
Comment (ashilay) [inline src/mqtt-client/client.js]: This let is used only inside publishDeviceInfoEvent() so can be moved also.

File: src/mqtt-client/client.js [hunk 8]
```diff
@@ -25,93 +26,59 @@ const client = mqtt.connect({
...
-                device: deviceName,
-                value: rawMessage.toString()
-            };
-        }
-        input.write(event);
-        debug(`Added to input stream event: '${JSON.stringify(event)}'`);
-    });
+function onMqttMessage(topic, rawMessage) {
+    debug(`MQTT >>. Got message: topic '${topic}', message: '${rawMessage.toString()}'`);
+    const event = convertMqttMessageToEvent(topic, rawMessage);
```
Comment (ashilay) [inline src/mqtt-client/client.js]: const {event} = convertMqttMessageToEvent(topic, rawMessage);
to get rid from event.event

File: src/mqtt-client/client.js [hunk 9]
```diff
@@ -25,93 +26,59 @@ const client = mqtt.connect({
...
+    input.write(event);
+    debug(`Added to input stream event: '${JSON.stringify(event)}'`);
 }
 
-function publishDeviceInfoEvent (device) {
-    let timeoutTimer;
-
-    if (device) {
-        publishingQueue.add(device);
+function onInputEvent(event) {
```
Comment (ashilay) [inline src/mqtt-client/client.js]: (event)

File: src/mqtt-client/client.js (hunk 8 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: as you can see bottom
input.write(event) is used !fixed !resolve

File: src/mqtt-client/client.js (hunk 9 above)
Comment (borys-ihnatyev) [inline src/mqtt-client/client.js]: as you can see bottom there is not only event property is used,  
for instance: JSON.stringify(event) !fix

**Pull request merged**